ENG-19266: Adds new modules for multinode PyTorch RDMA (#727)
opendatahub-io/opendatahub-documentation: modules/configuring-a-multi-node-pytorch-training-job-for-rdma.adoc

[source,subs="+quotes"]
----
apiVersion: mellanox.com/v1alpha1
kind: NicClusterPolicy
spec:
rdmaSharedDevicePlugin:
  config: |
    {
      "configList": [
        {
          "resourceName": "rdma_shared_device_eth",
          "rdmaHcaMax": 63,
          "selectors": {
            "ifNames": ["ens8f0np0"]
          }
        }
      ]
    }
----
+
In this example `NicClusterPolicy` resource, the `resourceName` value is `rdma_shared_device_eth`. 

.. Edit the `PyTorchJob` resource YAML code to add the following text:
+
.Example requests and limits for the network interface resources
[source,subs="+quotes"]
----
            resources:
              limits:
                nvidia.com/gpu: "1"
                rdma/rdma_shared_device_eth: "1"
              requests:
                nvidia.com/gpu: "1"
                rdma/rdma_shared_device_eth: "1"
----

.. In the `limits` and `requests` sections, replace the resource name with the resource name from your `NicClusterPolicy` resource (in this example, `rdma_shared_device_eth`).

.. Replace the specified value `1` with the number that you require.
Ensure that the specified amount is available on your OpenShift cluster.

> . Repeat the above steps to make the same edits in the `Worker` section of the PyTorchJob YAML code.

Reviewer: `PyTorchJob` ?

endif::[]

.Procedure
. Log in to the OpenShift Console.

. Create a `PyTorchJob` resource, as follows:
.. In the *Administrator* perspective, click *Home -> Search*.
.. From the *Project* list, select your project.
.. Click the *Resources* list, and in the search field, start typing `PyTorchJob`.
.. Select *PyTorchJob*, and click *Create PyTorchJob*.
+
The *Create PyTorchJob* page opens, with some basic YAML code automatically added.

. Attach the high-speed network interface to the PyTorchJob pods, as follows:

.. Edit the `PyTorchJob` resource YAML code to include an annotation that adds the pod to an additional network, as shown in the following example:
+
.Example annotation to attach network interface to pod
[source,subs="+quotes"]
----
spec:
  pytorchReplicaSpecs:
    Master:
      replicas: 1
      restartPolicy: OnFailure
      template:
        metadata:
          annotations:
            k8s.v1.cni.cncf.io/networks: "example-net"
----
.. Replace the example network name `example-net` with the appropriate value for your configuration.

. Configure the job to use NVIDIA Collective Communications Library (NCCL) interfaces, as follows:

> .. Edit the `PyTorchJob` resource YAML code to add the following text:

Reviewer 2: ```suggestion
.. Edit the `PyTorchJob` resource YAML code to add the following environment variables:
```

+
.Example environment variables
[source,subs="+quotes"]
----
        spec:
          containers:
          - command:
            - /bin/bash
            - -c
            - "your container command"
            env:
            - name: NCCL_SOCKET_IFNAME
              value: "net1"
            - name: NCCL_IB_HCA
              value: "mlx5_1"
----

> .. Replace the example container command and example environment-variable values with the appropriate values for your configuration:

Reviewer 2: ```suggestion
.. Replace the example environment-variable values with the appropriate values for your configuration:
```

> ... Replace `"your container command"` with the command to be run by your training job.

Reviewer 2: I think this can be removed as is self explanatory.
```suggestion
... 
```

[source,subs="+quotes"]
----
apiVersion: mellanox.com/v1alpha1
kind: NicClusterPolicy
spec:
rdmaSharedDevicePlugin:
  config: |
    {
      "configList": [
        {
          "resourceName": "rdma_shared_device_eth",
          "rdmaHcaMax": 63,
          "selectors": {
            "ifNames": ["ens8f0np0"]
          }
        }
      ]
    }
----
+
In this example `NicClusterPolicy` resource, the `resourceName` value is `rdma_shared_device_eth`. 

.. Edit the `PyTorchJob` resource YAML code to add the following text:
+
.Example requests and limits for the network interface resources
[source,subs="+quotes"]
----
            resources:
              limits:
                nvidia.com/gpu: "1"
                rdma/rdma_shared_device_eth: "1"
              requests:
                nvidia.com/gpu: "1"
                rdma/rdma_shared_device_eth: "1"
----

.. In the `limits` and `requests` sections, replace the resource name with the resource name from your `NicClusterPolicy` resource (in this example, `rdma_shared_device_eth`).

.. Replace the specified value `1` with the number that you require.
Ensure that the specified amount is available on your OpenShift cluster.

> . Specify the details of the nodes that will run the training job, as follows:

Reviewer 2: `Point 7. Specify the details of the nodes that will run the training job, as follows`
I think this section should be removed. Providing the entire example yaml later shows the user that they need to specify master and worker, which they may already know if they have experience with PyTorchJobs - so I think this is enough here.
